### ISSUE TITLE
[Feature Request] Enhance Reactive Controller

### ISSUE THREAD

**Issue by XGHeaven** [lit/lit#2148]
The controller is an amazing tool to reuse logic between component. But I think more thing can improve.

## Features

- `hostUpdated`/`hostUpdate` should accept `changedProperties` as argument.

- More lifecycle method should added. Such as `firstUpdated`/`attributeChangedCallback`/`adoptedCallback`

- The controller can has own reactive properties that be used for shared logic.

For example:

```typescript
class HiddenController {
  @property({type: Boolean, reflect: true}) hidden?: boolean

  constructor(public host: ReactiveControllHost) {}

  hostConnected() {
    (this.host as HTMLElement).style.display = this.hidden ? 'none' : undefined;
  }

  hostUpdated(changedProperties) {
    if (changedProperties.has('hidden')) {
      (this.host as HTMLElement).style.display = this.hidden ? 'none' : undefined;
    }
  }
}
```

**Comment by sorvell** [comment]
Thanks for the feedback. We intentionally kept the ReactiveController API pretty minimal to start to see what people actually need. 

We've considered adding `changedProperties` to the API and are currently weighing the pros and cons. This would tend to couple controllers and their hosts a bit, which may or may not be good. Likewise we're thinking about reactive properties in the controller, but this can be done in a custom controller.

**Comment by justinfagnani** [comment]
The problem with adding `changedProperties` to the API is that we'd like for reactive controllers to work across different libraries and not all libraries even have a meaningful set of changed properties. Function rendering approaches in particular don't.

There were also difficulties in reconciling both host and controller changed properties when we prototype controllers with reactive properties. Host and controller properties can't be in the same map otherwise there'd be collisions.

**Comment by XGHeaven** [comment]
Thanks for your response.

The first one, I think controller is a successor of mixin. So `changedProperties` and `reactive property` is necessary for controller.

Let me show a more complex example based on `MouseMoveController`.

1. Add attribute `mouse-move-delta` to control how much mouse moved distance before update.
2. Add attribute `mouse-move-disabled` to control whether update when mouse moved.

Without above feature, you must to do some work for every component try to using `MouseMoveController`:

1. You need to add `mouseMoveDelta`/`mouseMoveDisabled` as component property.
2. You need to keep `mouseMoveDelta`/`mouseMoveDisabled` sync with controller. (using `changedProperties` or property getter/setter or attributeChanged lifecycle)

```typescript
class SomeElement extends LitElement {
  @property({type: Number, reflect: true, attribute: 'mouse-move-delta'}) mouseMoveDelta?: number
  @property({type: Boolean, reflect: true, attribute: 'mouse-move-disabled'}) mouseMoveDisabled?: boolean

  mouse = new MouseMoveController(this)

  updated(changedProperties) {
    if (changedProperties.has('mouseMoveDelta')) {
      this.mouse.updateDelta(this.mouseMoveDelta)
    }
    if (changedProperties.has('mouseMoveDisabled')) {
      this.mouse.updateDisabled(this.mouseMoveDisabled)
    }
  }

  render() {
    // blahblah
  }
}

class MouseMoveController {
  delta = 200
  // extends docs demo https://lit.dev/docs/composition/controllers/#external-inputs
  updateDelta(delta) {
    this.delta = delta
  }

  updateDisabled(disabled) {
    if (disabled) {
      // remove listener
    } else {
      // add listener
    }
  }
}
```

This is a boring and time wasting work if I have many components. And also leads to an increase in error rates.
If above features can added, All of code can bundled in controller. Component need not to do anything else to use these capabilities.

The second, how do you think about more lifecycle should added?

**Comment by justinfagnani** [comment]
> The first one, I think controller is a successor of mixin. So changedProperties and reactive property is necessary for controller.

Controllers are _not_ successors to mixins. Controllers are a "has-a" form of composition that augments mixin's "is-a" form of composition. Controllers cannot add attributes, instance methods or properties (including reactive properties), and you can have multiple controller instances per class.

_Any_ instance members like the properties you have in your example have to be added by the element class, not a controller. Even if a controller had access to the host's `changedProperties`, the element class would still be responsible for declaring them. That puts a blurry coupling between host and controller.

See https://lit.dev/docs/composition/overview/#controllers-and-mixins for our current attempt to explain the different capabilities and use cases for controllers vs mixins.

Your example here is just better served by mixins, which can add reactive properties and access `changedProperties`. If you wished to separate the "is-a" and "has-a" parts for some reason, you could still do that:

```ts
const MouseMove = (Base: Constructor<LitElement) => class extends Base {
  @property({type: Number, reflect: true, attribute: 'mouse-move-delta'})
  mouseMoveDelta?: number;

  @property({type: Boolean, reflect: true, attribute: 'mouse-move-disabled'})
  mouseMoveDisabled?: boolean;

  mouse = new MouseMoveController(this);

  updated(changedProperties) {
    super.updated(changedProperties);
    if (changedProperties.has('mouseMoveDelta')) {
      this.mouse.updateDelta(this.mouseMoveDelta);
    }
    if (changedProperties.has('mouseMoveDisabled')) {
      this.mouse.updateDisabled(this.mouseMoveDisabled);
    }
  }
}

class MouseMoveController {
  delta = 200;
  // extends docs demo https://lit.dev/docs/composition/controllers/#external-inputs
  updateDelta(delta) {
    this.delta = delta;
  }

  updateDisabled(disabled) {
    if (disabled) {
      // remove listener
    } else {
      // add listener
    }
  }
}

class SomeElement extends MouseMove(LitElement) {
  render() {
    // blahblah
  }
}
```

**Comment by XGHeaven** [comment]
> Controllers are not successors to mixins. Controllers are a "has-a" form of composition that augments mixin's "is-a" form of composition. Controllers cannot add attributes, instance methods or properties (including reactive properties), and you can have multiple controller instances per class.

I understand what's the different between mixin and controller. But it's hard to beginners because Mixin can do anything that controller can do. There is no distinct line between mixin and controller to choose how to use them.
And Mixin has many problems for modern web development that are not recommended to use. For example, bad performance, prototype pollution, no type friendly and others.
So my opinion is deprecate mixin and enhance controller.

> Any instance members like the properties you have in your example have to be added by the element class, not a controller. Even if a controller had access to the host's changedProperties, the element class would still be responsible for declaring them. That puts a blurry coupling between host and controller.

I agree with you that would be increase coupling between host and controller indeed. But It's hard to maintain if I separate properties from controller. For example. I don't know what's usage of `mouseMoveDelta` in SomeElement if I only read `SomeElement` class code. I have to read all controllers of SomeElement if I want to find where use `mouseMoveDelta` property. It's so trouble.


> Your example here is just better served by mixins, which can add reactive properties and access changedProperties. If you wished to separate the "is-a" and "has-a" parts for some reason, you could still do that:
> (Example)

Yes, That's that I'm doing now. 😆  It's works but ugly.

**Comment by edwardwilson** [comment]
I would also like to see controllers have reactive properties. I think a `has-a` principle of a controller creating properties makes sense - I know these can be set on the component and passed to the controller but that becomes repetitive. I'm not sure this is the best example but a theme controller comes to mind; it might want a theme-name attribute. I would want to apply the controller and not worry about having to set up attrs etc

**Comment by justinfagnani** [comment]
@edwardwilson a utility that adds properties or attributes to the host should be a mixin.

**Comment by WickyNilliams** [comment]
Agreed about keeping controllers as generic as possible re: `changedProperties`. But any reason not to have a `firstUpdated` hook? I imagine every framework has ability to do one time setup? Of course, I can trivially implement it in a controller with a flag, but i think that obscures intent

**Comment by mbacalan** [comment]
> @edwardwilson a utility that adds properties or attributes to the host should be a mixin.

Excuse my lack of knowledge if there is a better way to do this, but using multiple mixins can get quite ugly and complicated real fast. For example, we have an in-house library where we want to share certain properties/attributes between components. Such as appearance, size, width and some validation. The resulting class is something like:

```js
export default class MyInput =>
  extends appearance(size(mask(vEmail(vMinlength(vMaxlength(formInput(wFull(LitElement)))))))) {
    // ...
  }
```

Wouldn't it be better if we could have these as controllers that are initialized in any host we want to use them with? We are currently unable to do this because we can't add reactive properties to the host from the controller.

**Comment by justinfagnani** [comment]
That is indeed no ideal. Instead, the `@watch` directive could just add metadata about which property is being watched and a single override of `update()` dispatches to the watch methods. This is how things will have to work for standard decorators, since they do not get access to the class. It would also be faster even for experimental decorators.

I think that changed property names in `update()` are going to be a bit of a legacy pattern once signals are added to JavaScript. Rather than coordinate on a string name and only be able to respond to reactive properties, we'll be able to make reactive properties be signals and things like these watch methods be called whenever any signal they use changes. That'll be more general, more ergonomic, and more robust to property name changes.

So I'd like to not spread changed properties more before that future is here, so we have less to undo.

**Comment by theacodes** [comment]
Fair enough! & yeah, things will hopefully get easier once we have Signals.

In the meantime, I modified `@watch` to use a mixin for the behavior and just have the decorator record the metadata. If anyone is curious, that's [here](https://gist.github.com/theacodes/a5ee193d7ca7a80d54a58114e1f5a6f7). This pattern can probably be adapted to a lot of the use cases mentioned here, though I agree that having to manually add the mixin is a little less ergonomic.

**Comment by justinfagnani** [comment]
Thanks for the example code @theacodes. I also filed an [issue](https://github.com/shoelace-style/shoelace/issues/1990) with Shoelace (with some example code too, fyi), so hopefully they can improve their implementation as well.

**Comment by theacodes** [comment]
Awesome. It looks like since there are many ways to work around this or otherwise accomplish the intent and the controller interface is intentionally minimal, it might make sense to close this?
